client/src/test/java/org/eclipse/hono/client/impl/HonoConnectionImplTest.java

```diff
@@ -648,6 +648,47 @@ public void testCreateReceiverFailsOnTimeout(final TestContext ctx) {
         verify(remoteCloseHook, never()).handle(anyString());
     }
 
+    /**
+     * Verifies that the attempt to create a receiver fails with a
+     * {@code ServerErrorException} if the connection gets disconnected
+     * before the remote peer has sent its attach frame. It is verified
+     * that this is done before the link establishment timeout.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testCreateReceiverFailsOnDisconnectBeforeOpen(final TestContext ctx) {
+
+        final long linkEstablishmentTimeout = 444L; // choose a distinct value here
+        props.setLinkEstablishmentTimeout(linkEstablishmentTimeout);
+        // don't run linkEstablishmentTimeout timer handler
+        when(vertx.setTimer(eq(linkEstablishmentTimeout), any(Handler.class))).thenAnswer(invocation -> 0L);
+
+        // GIVEN an established connection
+        final Async connectAttempt = ctx.async();
+        honoConnection.connect().setHandler(ctx.asyncAssertSuccess(ok -> connectAttempt.complete()));
+        connectAttempt.await();
+        final Source source = mock(Source.class);
+        when(source.getAddress()).thenReturn("source/address");
+        final ProtonReceiver receiver = mock(ProtonReceiver.class);
+        when(receiver.isOpen()).thenReturn(Boolean.TRUE);
+        when(receiver.getSource()).thenReturn(source);
+        when(receiver.getRemoteSource()).thenReturn(source);
+        when(con.createReceiver(anyString())).thenReturn(receiver);
+
+        // WHEN creating a receiver link with a close hook
+        final Handler<String> remoteCloseHook = mock(Handler.class);
+
+        final Future<ProtonReceiver> result = honoConnection.createReceiver("source", ProtonQoS.AT_LEAST_ONCE,
+                mock(ProtonMessageHandler.class), remoteCloseHook);
```

Comment:
IMHO you also nee to verify that at this point the result is not completed/failed yet ...